### DISCUSSION TITLE
Set VRF as well as interface in curl option

### DISCUSSION THREAD

**Discussion by Avi-Roy1** [curl/curl#9913] [Q&A]
Hello All,

Recently I had a requirement where I need to use a specific interface for outbound traffic (CURL) in a Linux system. My application needs to run in default vrf as it's communicating with other processes internally that are running on another HW. As I'm communicating outside I need to mention vrf too (MGMT). Now the issue is curl supports only one option to be set, it's either vrf or an interface via CURLOPT_INTERFACE.
Could someone please guide me to set both of these options?

Thanks,
Avi

**Comment by bagder** [comment]
I don't understand. The `CURLOPT_INTERFACE` is one curl option. Which is the other?

**Reply by Avi-Roy1** [reply]
Hello,

That's the problem! When I checked the documentation CURLOPT_INTERFACE could either set an interface or vrf but **I need to set both of them**. So is there any way I could achieve that?

**Reply by Avi-Roy1** [reply]
please let me know if we can set both **vrf-blue** and **interface-LO** in the curl option.

**Reply by jay** [reply]
Though I am not using VRFs from what I understand you are supposed to assign them to an interface. And then, presumably, you'd select that interface with libcurl. Or am I missing something?

**Reply by Avi-Roy1** [reply]
I would like to set both vrf as well as an interface via the curl option, but it seems only one could be set at a time.
ref: https://github.com/curl/curl/pull/2024/files
Do we have any other way to set these options(vrf as well interface) together? or could we adapt the curl code to set both of these options?

**Reply by bagder** [reply]
How do you set vrf for a socket/connection?

**Reply by Avi-Roy1** [reply]
We used CURLOPT_INTERFACE option to set vrf in curl. 
socket option could be set via setsockopt option.
eg: setsockopt(sd, SOL_SOCKET, SO_BINDTODEVICE, dev, strlen(dev)+1);
ref: https://www.kernel.org/doc/Documentation/networking/vrf.txt

**Reply by bagder** [reply]
As you can only bind the socket to a single device/interface, does this not answer your question? It seems you are asking for something the underlying API does not allow for.

So we are back to @jay's suggestion: "you are supposed to assign them to an interface and then use that"

**Comment by dfandrich** [comment]
IP only lets you send over one interface at a time. If you want to send from
more than one at a time, you'll need to set up a bridge interface or something
more sophisticated. What exactly are you trying to accomplish?

**Reply by Avi-Roy1** [reply]
I have a process that runs in default vrf which uses libcurl. I need to set the source interface for outgoing traffic and also need to specify the vrf (eg: vrf-blue) via which the traffic should go. From the curl documentation, I could see we could use option CURLOPT_INTERFACE for setting either vrf or interface but not both. So, are there any ways to achieve this?

**Comment by Avi-Roy1** [comment]
because there is no way we could set vrf as well as the interface together, I am sending the curl command to the MGMT interface via "IP vrf exec <curl -T <URL>  --interface <interface name>> prefix to the shell and it worked. I had to use curl binary here.
